WalletWasabi/WabiSabi/Crypto/UnboundedChannel.cs

```diff
@@ -0,0 +1,37 @@
+using System.Threading;
+using System.Threading.Tasks;
+using System.Collections.Concurrent;
+
+namespace WalletWasabi.WabiSabi.Crypto
+{
+	public class UnboundedChannel<T>
+	{
+		private readonly ConcurrentQueue<T> _queue = new();
+		private readonly ConcurrentQueue<TaskCompletionSource<T>> _waitingQueue = new();
+
+		public void Send(T item)
```

Comment:
Hard to figure out what is happening from the naming. If we will go with this I suggest renaming things here.